File: examples/bluesolar_angles/main2.cpp
#include <iomanip>
#include <iostream>
#include <vector>
#include <map>
#include <tuple>
#include <cmath>

#include "TrackerArmature2A.h"
#include "TrackerTarget.h"
#include "Ray.h"
#include "Transform.h"
#include "ElevationAngleKM.h"
#include "HourAngleKM.h"
#include "BluesolarHelpers.h"

int main(int argc, char *argv[])
{
    // Global coordinate system x > 0 towards East, y > 0 towards North, z > 0 towards Zenith
    // Positive angles in the counterclockwise direction according to the direction of the rotation axis

    // Declaracion of variables to store the input values to the program passed as line arguments
    double xHeliostat, yHeliostat, zHeliostat; // Heliostat position on the field.
    double xAimingPoint, yAimingPoint, zAimingPoint; // Aiming point.
    double sunAzimuth, sunElevation; // Azimuth and elevation of the Sun in degrees.

    // Conversion and validation of line argument values from string to doubles
    if ( !ConvertAndValidateArguments(argc, argv, xHeliostat, yHeliostat, zHeliostat, 
                                     xAimingPoint, yAimingPoint, zAimingPoint, 
                                     sunAzimuth, sunElevation) ) return 1; // Return with error code

    // Defining the target
    TrackerTarget* target = new TrackerTarget;
    target->aimingPoint = vec3d(xAimingPoint, yAimingPoint, zAimingPoint);

    // Setting up the tracker armature for the BlueSolar heliostat
    TrackerArmature2A armature;
    BluesolarHeliostatArmature(armature);

    // Moving the heliostat armature to the location defined by the user
    Transform heliostatLocation = Transform::translate(xHeliostat, yHeliostat, zHeliostat); // actual location of a heliostat in the bluesolar field

    // Defining the sun vector in the global coordinate system from sun angles (azimuth and elevation) in degrees provided by the user.
    // Azimuth = 0 at North and > 0 towards East. Elevation = 0 at the horizontal plane and > 0 towards zentihg
    double sun_azimuth = sunAzimuth * gcf::degree;
    double sun_elevation = sunElevation * gcf::degree;
    vec3d sunVector(cos(sun_elevation)*sin(sun_azimuth), cos(sun_elevation)*cos(sun_azimuth),sin(sun_elevation));

    // Computing the angles the heliostat armature should turn to reflect the direct solar radiation to the target
    armature.update(heliostatLocation, sunVector, target);

    double elevation_angle = (target->angles.x) * gcf::degree; // radians
    double hour_angle = (target->angles.y) * gcf::degree; // radians

     // Setting up the elevation and hour angle kinematic models for the BlueSolar heliostat
    ElevationAngleKM elevation_angle_km = CreateBluesolarElevationAngleKM();
    HourAngleKM hour_angle_km = CreateBluesolarHourAngleKM();

    double length_elevation_angle_actuator = elevation_angle_km.getActuatorLengthFromElevationAngle(elevation_angle);
    double length_hour_angle_actuator = hour_angle_km.getActuatorLengthFromHourAngle(hour_angle);

    // Printing the results
    std::cout << target->angles << ", " << length_elevation_angle_actuator << ", " << length_hour_angle_actuator << std::endl;

    delete target;
    return 0; // Return with success code
}

File: examples/bluesolar_angles_python/main2.py
# main2.py
import sys
import os
from math import cos, sin, radians

# Ensure Python can find the compiled module
sys.path.append('C:/Users/manue_6t240gh/Dropbox/OpenSource/heliostat_tracking/build/Release')

import numpy as np
import heliostat_tracking_module as ht
from BluesolarHelpers import create_bluesolar_elevation_angle_km, create_bluesolar_hour_angle_km

def main():
    if len(sys.argv) != 9:
        print("Usage: main2.py xHeliostat yHeliostat zHeliostat xAimingPoint yAimingPoint zAimingPoint sunAzimuth sunElevation")
        return

    xHeliostat, yHeliostat, zHeliostat = map(float, sys.argv[1:4])
    xAimingPoint, yAimingPoint, zAimingPoint = map(float, sys.argv[4:7])
    sunAzimuth, sunElevation = map(float, sys.argv[7:9])

    # Create TrackerTarget instance
    target = ht.TrackerTarget()
    target.aiming_point = ht.vec3d(xAimingPoint, yAimingPoint, zAimingPoint)

    # Create and configure TrackerArmature2A instance
    armature = ht.TrackerArmature2A()
    armature.set_primary_shift(ht.vec3d(0.0, 0.0, 2.415))
    armature.set_primary_axis(ht.vec3d(-1.0, 0.0, 0.0))
    armature.set_primary_angles(ht.vec2d(0.0, 90.0))
    armature.set_secondary_shift(ht.vec3d(0.0, -0.0816, 0.0))
    armature.set_secondary_axis(ht.vec3d(0.0, 0.0, -1.0))
    armature.set_secondary_angles(ht.vec2d(-70.0, 55.0))
    armature.set_facet_shift(ht.vec3d(0.0, -0.105, 0.035))
    armature.set_facet_normal(ht.vec3d(0.0, -1.0, 0.0))
    armature.set_default_angles(ht.vec2d(0.0, 0.0))

    # Define the heliostat location transform
    heliostat_location = ht.Transform([
        [1.0, 0.0, 0.0, xHeliostat],
        [0.0, 1.0, 0.0, yHeliostat],
        [0.0, 0.0, 1.0, zHeliostat],
        [0.0, 0.0, 0.0, 1.0]
    ])

    # Define the sun vector from sun angles
    sun_azimuth = radians(sunAzimuth)  # Convert to radians
    sun_elevation = radians(sunElevation)  # Convert to radians
    sun_vector = ht.vec3d(cos(sun_elevation) * sin(sun_azimuth), cos(sun_elevation) * cos(sun_azimuth), sin(sun_elevation))

    # Update the armature to compute the angles
    armature.update(heliostat_location, sun_vector, target)

    # Get the resulting angles
    elevation_angle = target.angles.x
    hour_angle = target.angles.y

    # Initialize kinematic models using BlueSolarHelpers
    elevation_angle_km = create_bluesolar_elevation_angle_km()
    hour_angle_km = create_bluesolar_hour_angle_km()

    # Compute actuator lengths
    length_elevation_angle_actuator = elevation_angle_km.get_actuator_length_from_elevation_angle(radians(elevation_angle))
    length_hour_angle_actuator = hour_angle_km.get_actuator_length_from_hour_angle(radians(hour_angle))

    # Print the results
    print(f"Angles: ({target.angles.x}, {target.angles.y}), Elevation Actuator Length: {length_elevation_angle_actuator}, Hour Actuator Length: {length_hour_angle_actuator}")

if __name__ == "__main__":
    main()